### PR TITLE
layered: MO save transitive dependencies when comparing ports by model order.

### LINKED ISSUE #918: ModelOrderPortComparator seem to not fulfil the comparator contract
For some of our models we get this exception thrown. It looks like the `ModelOrderPortComparator` implementation is bogus.
I can't provide the test data yet, but will do so as soon as I reduced our huge model to a small reproducible example.

```
stack trace: java.lang.IllegalArgumentException: Comparison method violates its general contract!
	at java.base/java.util.TimSort.mergeLo(TimSort.java:781)
	at java.base/java.util.TimSort.mergeAt(TimSort.java:518)
	at java.base/java.util.TimSort.mergeCollapse(TimSort.java:448)
	at java.base/java.util.TimSort.sort(TimSort.java:245)
	at java.base/java.util.Arrays.sort(Arrays.java:1515)
	at java.base/java.util.ArrayList.sort(ArrayList.java:1750)
	at java.base/java.util.Collections.sort(Collections.java:179)
	at org.eclipse.elk.alg.layered.intermediate.SortByInputModelProcessor.process(SortByInputModelProcessor.java:69)
	at org.eclipse.elk.alg.layered.intermediate.SortByInputModelProcessor.process(SortByInputModelProcessor.java:1)
	at org.eclipse.elk.alg.layered.ElkLayered.layout(ElkLayered.java:606)
	at org.eclipse.elk.alg.layered.ElkLayered.doLayout(ElkLayered.java:152)
	at org.eclipse.elk.alg.layered.LayeredLayoutProvider.layout(LayeredLayoutProvider.java:56)
	at org.eclipse.elk.core.RecursiveGraphLayoutEngine.executeAlgorithm(RecursiveGraphLayoutEngine.java:250)
	at org.eclipse.elk.core.RecursiveGraphLayoutEngine.layoutRecursively(RecursiveGraphLayoutEngine.java:218)
	at org.eclipse.elk.core.RecursiveGraphLayoutEngine.layout(RecursiveGraphLayoutEngine.java:96)
	at org.eclipse.elk.core.RecursiveGraphLayoutEngine.layout(RecursiveGraphLayoutEngine.java:72)
	at io.typefox.elkserver.StdioElkServer.run(StdioElkServer.java:27)
	at io.typefox.elkserver.AbstractElkServer.start(AbstractElkServer.java:35)
	at io.typefox.elkserver.ElkServerApplication.main(ElkServerApplication.java:25)
```

I also noticed that the implementation is not that performant. The bottleneck in my case is due to many `HashSet` reads.

### LINKED ISSUE #930: Port Model Order not respected for not connected ports
See discussion on gitter
https://matrix.to/#/!hFJKlkXrdolGTZEIdx:gitter.im/$IEzrWXP6ZmWXwyTEmiY4l3evptDsfR1X-74Jo2yNvso?via=gitter.im&via=matrix.org

### PR DESCRIPTION
**Pull request by soerendomroes**
Fixed #918 
Fixes #930

### CODE REVIEW

**Comment by soerendomroes** [comment]
@Eddykasp maybe you can have a look at this.

File: plugins/org.eclipse.elk.alg.layered/src/org/eclipse/elk/alg/layered/intermediate/preserveorder/ModelOrderPortComparator.java [hunk 1]
```diff
@@ -43,6 +46,15 @@ public class ModelOrderPortComparator implements Comparator<LPort> {
     private LNode[] previousLayer;
     
     private OrderingStrategy strategy;
+    
+    /**
+     * Each node has an entry of nodes for which it is bigger.
```
Comment (Eddykasp) [inline plugins/org.eclipse.elk.alg.layered/src/org/eclipse/elk/alg/layered/intermediate/preserveorder/ModelOrderPortComparator.java:51]: shouldn't it say "each port has an entry of ports for which it is bigger"?

File: plugins/org.eclipse.elk.alg.layered/src/org/eclipse/elk/alg/layered/intermediate/preserveorder/ModelOrderPortComparator.java [hunk 2]
```diff
@@ -43,6 +46,15 @@ public class ModelOrderPortComparator implements Comparator<LPort> {
...
+     */
+    private HashMap<LPort, HashSet<LPort>> biggerThan = new HashMap<>();
+    /**
+     * Each node has an entry of nodes for which it is smaller.
```
Comment (Eddykasp) [inline plugins/org.eclipse.elk.alg.layered/src/org/eclipse/elk/alg/layered/intermediate/preserveorder/ModelOrderPortComparator.java:55]: see above comment

File: plugins/org.eclipse.elk.alg.layered/src/org/eclipse/elk/alg/layered/intermediate/preserveorder/ModelOrderPortComparator.java [hunk 3]
```diff
@@ -180,10 +283,44 @@ public int compare(final LPort p1, final LPort p2) {
...
+        if (p1.hasProperty(InternalProperties.MODEL_ORDER)
+                && p2.hasProperty(InternalProperties.MODEL_ORDER)) {
             return Integer.compare(p1.getProperty(InternalProperties.MODEL_ORDER),
                     p2.getProperty(InternalProperties.MODEL_ORDER));
         }
         return 0;
     }
+    
+    private void updateBiggerAndSmallerAssociations(final LPort bigger, final LPort smaller) {
+        HashSet<LPort> biggerNodeBiggerThan = biggerThan.get(bigger);
```
Comment (Eddykasp) [inline plugins/org.eclipse.elk.alg.layered/src/org/eclipse/elk/alg/layered/intermediate/preserveorder/ModelOrderPortComparator.java:295]: again, is this comparing nodes or ports?

**Review by Eddykasp — COMMENTED** [review]
It's not totally obvious here what bigger and smaller with respect ports (nodes?) means here. Is it simply the ordering around the node i.e. top-left corner -> top-right corner and top-left corner -> bottom-left corner, etc.?

File: test/org.eclipse.elk.alg.layered.test/src/org/eclipse/elk/alg/layered/intermediate/greedyswitch/TestGraphCreator.java [hunk 4]
```diff
@@ -721,7 +721,7 @@ public LGraph getNodesInDifferentLayoutUnitsPreventSwitch() {
         LNode[] leftNodes = addNodesToLayer(2, layers[0]);
         LNode[] rightNodes = addNodesToLayer(3, layers[1]);
 
-        eastWestEdgeFromTo(leftNodes[0], rightNodes[1]);
+        eastWestEdgeFromTo(leftNodes[1], rightNodes[0]);
```
Comment (Eddykasp) [inline test/org.eclipse.elk.alg.layered.test/src/org/eclipse/elk/alg/layered/intermediate/greedyswitch/TestGraphCreator.java:724]: this makes sense to me

**Comment by dhuebner** [comment]
@soerendomroes 
Thanks for solving the issues! 
Do you have an idea on how the changes impact the performance?

**Comment by soerendomroes** [comment]
Yes, probably negatively since the algorithm needs even more Map reads.

**Pull request merged**